kore/src/Kore/Attribute/Concat.hs

```diff
@@ -0,0 +1,94 @@
+{-|
+Module      : Kore.Attribute.Concat
+Description : Concat attribute
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+Maintainer  : thomas.tuegel@runtimeverification.com
+
+-}
+
+{-# LANGUAGE Strict #-}
+
+module Kore.Attribute.Concat
+    ( Concat (..)
+    , mergeConcat, toConcat, fromConcat
+    , concatId, concatSymbol, concatAttribute
+    ) where
+
+import Prelude.Kore
+
+import Data.Default
+import qualified Generics.SOP as SOP
+
+import Kore.Attribute.Parser
+import Kore.Debug
+import Kore.Unparser
+
+-- | @Concat@ represents the @concat@ attribute.
+newtype Concat symbol = Concat { getConcat :: Maybe symbol }
+    deriving (Generic, Eq, Ord, Show)
+    deriving (Foldable, Functor, Traversable)
+    deriving anyclass (Hashable)
+    deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
+    deriving anyclass (Debug, Diff)
+
+mergeConcat :: (Eq symbol , Unparse symbol)
+    => Concat symbol -> Concat symbol -> Concat symbol
+mergeConcat (Concat Nothing) b = b
+mergeConcat a (Concat Nothing) = a
+mergeConcat a@(Concat (Just aSymbol)) (Concat (Just bSymbol))
+      | aSymbol == bSymbol = a
+      | otherwise = error $
+        "Concat symbol mismatch error! Found both "
+        ++ unparseToString aSymbol
+        ++ " and "
+        ++ unparseToString bSymbol
+        ++ " inside term. This is a bug in the Haskell backend."
+
+instance Default (Concat symbol) where
+    def = Concat { getConcat = Nothing }
+
+instance NFData symbol => NFData (Concat symbol)
+
+instance ParseAttributes (Concat SymbolOrAlias) where
+    parseAttribute = withApplication' parseApplication
+      where
+        parseApplication params args Concat { getConcat }
+          | Just _ <- getConcat = failDuplicate'
+          | otherwise = do
+            getZeroParams params
+            arg <- getOneArgument args
+            symbol <- getSymbolOrAlias arg
+            return Concat { getConcat = Just symbol }
+        withApplication' = withApplication concatId
+        failDuplicate' = failDuplicate concatId
+
+instance From (Concat SymbolOrAlias) Attributes where
+    from =
+        maybe def toAttribute . getConcat
+      where
+        toAttribute = from @AttributePattern . concatAttribute
+
+toConcat :: symbol -> Concat symbol
+toConcat sym = Concat { getConcat = Just sym }
+
+fromConcat :: Concat symbol -> symbol
+fromConcat Concat {getConcat = Just sym} = sym
+fromConcat _ = error "There is no concat symbol to extract"
```

Comment:
Why is this partial function necessary?

Reply:
@ana-pantilie The partial functions are used in unparsing and externalisation (more precisely, they are used in `asTermLike`). Unfortunately, we cannot refactor the code into not using them as the symbols contained within are required to be there for these actions. If they aren't, something, somewhere has gone wrong and we need to error.
The code is however written in such a way as to not ask for these symbols until they are absolutely necessary, so there shouldn't be instances of the code erroring without needing to (e.g. we don't call `fromConcat` or `fromElement` when externalizing/unparsing an empty map).

Reply:
Ok, but this sounds to me like the representation of those symbols in those specific contexts isn't right. I believe this is caused by the changes to `InternalAc`. I assume that, for some reason, we decided that fixing this isn't practical for the scope of this PR, right?

Reply:
I mean, I want to be sure that we consciously made this decision.

Reply:
I'm not sure I understand what you mean. Could you, please, elaborate more on that?

Reply:
Resolved during the meeting: this decision was discussed and was decided to be the simplest solution.